drgn_tools/cpuinfo.py

```diff
@@ -0,0 +1,615 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helper to view cpuinfo data
+"""
+import argparse
+from typing import Any
+from typing import Dict
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.linux.percpu import per_cpu
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.table import print_dictionary
+
+
+def x86_get_cpu_info(prog: Program) -> Dict[str, Any]:
+    """
+    Helper to get cpuinfo data for x86
+
+    :returns: a dictionary of the cpuinfo data
+    """
+    # TODO: Add aarch64 support.
+    # `struct cpuinfo_arm64` being fetched using 'boot_cpu_data' only has
+    # register values, while the struct fetched using 'cpu_data' has all values
+    #  zero/null
+
+    cpus = int(prog["nr_cpu_ids"])
+
+    if "cpu_data" in prog:
+        cpuinfo_struct = prog["cpu_data"]
+    elif "boot_cpu_data" in prog:
+        cpuinfo_struct = prog["boot_cpu_data"]
+    else:
+        raise Exception(
+            "Failed to load CPU info: no cpuinfo struct found (tried 'cpu_data' and 'boot_cpu_data')"
+        )
+
+    cpu_vendor = cpuinfo_struct.x86_vendor_id.string_().decode("utf-8")
+    model_name = cpuinfo_struct.x86_model_id.string_().decode("utf-8")
+    cpu_family = int(cpuinfo_struct.x86)
+    cpus_numa0 = "0-" + str(cpus - 1)
+    microcode = hex(cpuinfo_struct.microcode)
+    cstates = int(prog["max_cstate"])
+
+    cpu_caps_bugs = cpuinfo_struct.x86_capability  # Contains cpu and bug flags
+
+    nbugints = 1
+    ncapints = len(cpu_caps_bugs) - nbugints
+
+    cpu_flags = prog["x86_cap_flags"]
+    bug_flags = prog["x86_bug_flags"]
+
+    cpu_flags_list = []
+    bug_flags_list = []
+
+    for i in range(ncapints):
+        flags = int(cpu_caps_bugs[i])
+        for j in range(32):
+            if (flags & 1) and cpu_flags[i * 32 + j]:
+                cpu_flags_list.append(
+                    cpu_flags[i * 32 + j].string_().decode("utf-8")
+                )
+            flags >>= 1
+
+    for i in range(nbugints):
+        flags = int(cpu_caps_bugs[i + ncapints])
+        for j in range(32):
+            if (flags & 1) and bug_flags[i * 32 + j]:
+                bug_flags_list.append(
+                    bug_flags[i * 32 + j].string_().decode("utf-8")
+                )
+            flags >>= 1
```

Comment:
I think we can simplify this substantially. A few notes:

1. Drgn has a `for_each_set_bit()` helper that will make the loops much nicer.
2. On the master branch of Linux (and indeed, OL9 RHCK), `NBUGINTS` is already 2. We can't hard-code that value. Instead, let's infer it from the size of `x86_bug_flags`. In fact... we don't really even need to infer those sizes directly!
3. Let's also use `read_()` here to avoid repeated re-reading of the array data.

```suggestion
    cap = cpuinfo_struct.x86_capability
    cap_flags = prog["x86_cap_flags"].read_()
    bug_flags = prog["x86_bug_flags"].read_()

    cpu_flags_list = []
    bug_flags_list = []

    for nr in for_each_set_bit(cap, 8 * sizeof(cap)):
        if nr < len(cap_flags) and cap_flags[nr]:
            cpu_flags_list.append(cap_flags[nr].string_().decode("utf-8"))
        elif nr >= len(cap_flags) and bug_flags[nr - len(cap_flags)]:
            bug_flags_list.append(bug_flags[nr - len(cap_flags)].string_().decode("utf-8"))
```